google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/WriterCache.java

```diff
@@ -0,0 +1,245 @@
+
+  // Maximum number of tables to hold in the cache, once the maxium exceeded, the cache will be
+  // evicted based on least recent used.
+  private static final int MAX_TABLE_ENTRY = 100;
+  private static final int MAX_WRITERS_PER_TABLE = 2;
+
+  private final BigQueryWriteClient stub;
+  private final SchemaCompact compact;
+
+  private WriterCache(BigQueryWriteClient stub, int maxTableEntry, SchemaCompact compact) {
+    this.stub = stub;
+    this.compact = compact;
+    writerCache =
+        new LRUCache<String, LRUCache<Descriptors.Descriptor, StreamWriter>>(maxTableEntry);
+  }
+
+  public static WriterCache getInstance() throws IOException {
+    if (instance == null) {
+      BigQueryWriteSettings stubSettings = BigQueryWriteSettings.newBuilder().build();
+      BigQueryWriteClient stub = BigQueryWriteClient.create(stubSettings);
+      instance = new WriterCache(stub, MAX_TABLE_ENTRY, SchemaCompact.getInstance());
+    }
+    return instance;
+  }
+
+  /** Returns a cache with custom stub used by test. */
+  @VisibleForTesting
+  public static WriterCache getTestInstance(
+      BigQueryWriteClient stub, int maxTableEntry, SchemaCompact compact) {
+    return new WriterCache(stub, maxTableEntry, compact);
+  }
+
+  /** Returns an entry with {@code StreamWriter} and expiration time in millis. */
+  private String CreateNewStream(String tableName) {
+    Stream.WriteStream stream =
+        Stream.WriteStream.newBuilder().setType(Stream.WriteStream.Type.COMMITTED).build();
+    stream =
+        stub.createWriteStream(
+            Storage.CreateWriteStreamRequest.newBuilder()
+                .setParent(tableName)
+                .setWriteStream(stream)
+                .build());
+    LOG.info("Created write stream:" + stream.getName());
+    return stream.getName();
+  }
+
+  StreamWriter CreateNewWriter(String streamName)
+      throws IllegalArgumentException, IOException, InterruptedException {
+    return StreamWriter.newBuilder(streamName)
+        .setChannelProvider(stub.getSettings().getTransportChannelProvider())
+        .setCredentialsProvider(stub.getSettings().getCredentialsProvider())
+        .setExecutorProvider(stub.getSettings().getExecutorProvider())
+        .build();
+  }
+  /**
+   * Gets a writer for a given table with a given user schema from global cache.
+   *
+   * @param tableName
+   * @param userSchema
+   * @return
+   * @throws Exception
+   */
+  public StreamWriter getTableWriter(String tableName, Descriptors.Descriptor userSchema)
+      throws IllegalArgumentException, IOException, InterruptedException {
+    Matcher matcher = tablePattern.matcher(tableName);
+    if (!matcher.matches()) {
+      throw new IllegalArgumentException("Invalid table name: " + tableName);
+    }
+
+    String streamName = null;
+    Boolean streamExpired = false;
+    StreamWriter writer = null;
+    LRUCache<Descriptors.Descriptor, StreamWriter> tableEntry = null;
+
+    synchronized (this) {
```

Comment:
I found this logic a bit hard to read and wonder if we could refactor into this:
``` synchronized (this) {
      tableEntry = writerCache.get(tableName);
      if (tableEntry != null) {
        writer = tableEntry.get(userSchema);
        if (writer != null) {
          if (writer.expired())
            writer.close();
        return writer;
        } else {
          compact.check(tableName, userSchema);
          streamName = CreateNewStream(tableName);
          writer = CreateNewWriter(streamName);
          tableEntry.put(userSchema, writer);
        }
      } else {
        compact.check(tableName, userSchema);
        streamName = CreateNewStream(tableName);
        writer = CreateNewWriter(streamName);
        tableEntry = new LRUCache<Descriptors.Descriptor, StreamWriter>(MAX_WRITERS_PER_TABLE);
        tableEntry.put(userSchema, writer);
        writerCache.put(tableName, tableEntry);
      }
    } ```

Reply:
Done.